src/systems/triggered_publisher/TriggeredPublisher.hh

```diff
@@ -210,31 +223,68 @@ namespace systems
       transport::Node::Publisher pub;
     };
 
+    /// \brief Class that holds necessary bits for each specified service output
+    private: struct ServiceOutputInfo
+    {
+      /// \brief Service name
+      std::string servName;
+
+      /// \brief Service request type
+      std::string reqType;
+
+      /// \brief Service response type
+      std::string repType;
+
+      /// \brief Service request message
+      std::string reqMsg;
+
+      /// \brief Service request timeout
+      unsigned int timeout;
+    };
+
     /// \brief List of InputMatchers
     private: std::vector<std::unique_ptr<InputMatcher>> matchers;
 
     /// \brief List of outputs
     private: std::vector<OutputInfo> outputInfo;
 
+    /// \brief List of service outputs
+    private: std::vector<ServiceOutputInfo> serviceOutputInfo;
+
     /// \brief Ignition communication node.
     private: transport::Node node;
 
     /// \brief Counter that tells the publisher how many times to publish
     private: std::size_t publishCount{0};
 
+    /// \brief Counter that tells how many times to call the service
+    private: std::size_t serviceCount{0};
```

Comment:
Do you expect this to ever be > 1? I'm trying to think when you would want to call a service more than once with the same information. Wouldn't they call it, then wait for timeout, and if it's unsuccessful, call it again? Publishing topics is "free," as in it just throws it onto the network, but calling a service multiple times means the handler has to do the same work multiple times, which may be redundant and unnecessary.

Reply:
--- For publishing---
So the `publishCount` increases by 1 every time a new msg is received in the callback. And there's a separate `DoWork` thread that uses this `publishCount` to determine when to start publishing and how many times. So if N number of input msg is received, N number of output msg will be published.

So I thought of using the same logic where `serviceCount` will be >1 if the [callback](https://github.com/gazebosim/gz-sim/blob/liam/add-serv-to-trig-pub/src/systems/triggered_publisher/TriggeredPublisher.cc#L668-L669) is triggered more than once. As in, if a msg is published to the topic that triggers the callback more than once. 

That being said, after contemplating for a while, I think you have a valid point. I don't think we need to "queue" up the service like the publisher. So I can change it so that the callback alerts the DoServiceWork thread to call the service. And if the callback is triggered again while the service is processing, I'll disregard the new service request. (This I believe should NOT happen unless a user decides to e.g. click a reset button multiple times hence publishing a msg that'll trigger the callback). Let me know your thoughts. Also, if any of this is confusing, I think looking over the code and talking in person might be more clear haha

Reply:
Oh I see. Hmm okay I think this is fine then.